URLSearchParams Implementation

etiennemartin opened the pull request:
This PR is a proposal for the implementation of the `URLSerachParams` class provided by NodeJS. You can find the documentation here: [Docs](https://nodejs.org/docs/latest/api/url.html#class-urlsearchparams).

This implementation is continuation of the work done to implement URL itself. It tries to implement as much of the NodeJS functionality as possible for goja.

The following is currently unsupported in the implementation:

- `ctor(iterable)`: Using function generators
- `sort()`:  Since the backing object is a url.URL which backs the data as a Map, we can't reliably sort the entries
- `[] operator`: Not sure if we are able to override this with goja.

If the general approach is something we are happy with, I will look at making sorting and ordering something built in. This would require more work since we wouldn't be able to rely on the backing object being a `url.URL` struct. Right now we use the `Encode()` of `url.URL` method in several places.

etiennemartin commented on url/url.go:
```diff
@@ -0,0 +1,331 @@
+	})
+
+	// search Params
+	defineURLAccessorProp(r, p, "searchParams", func(u *url.URL) interface{} {
```
Most of the code in this file was simply moved from `module.go`. I've only added this new method as part of the URL implementation (Along with tests).

etiennemartin commented:
The latest changes shows that a random change can affect the ordering of the url parameters. This results in flaky tests. I'll have to find a way to get this working with something consistent. My idea right now is to wrap a struct and maintain a separate list of ordered parameters.

dop251 commented:
Hi. I've just realised that in node the url implementation is actually written in javascript: https://github.com/nodejs/node/blob/main/lib/internal/url.js

I think for maximum compatibility the implementation should be derived from there, converted to Go of course.

I had a quick look at the code, one thing that strikes me is that they set the `search` property of the associated URL every time the query parameters are updated... Not quite sure why not make it a lazy update since `URL.search` is an accessor property.

As for the iteration, I guess goja needs to expose something that would allow doing `for ... of` from Go.

Not quite sure what you mean by `[] operator`.

etiennemartin commented:
> I think for maximum compatibility the implementation should be derived from there, converted to Go of course.

How important is the idea of maximum compatibility here? I'm just wondering since this would invalidate the implementation of `URL` and `URLSearchParams`. If we aim to be as close as possible it would suggest a re-write for the most part. My assumption here was to ensure that the implementation would behave the same, not exactly implemented the same. Just trying to align expectations moving forward.

> I had a quick look at the code, one thing that strikes me is that they set the search property of the associated URL every time the query parameters are updated... Not quite sure why not make it a lazy update since URL.search is an accessor property.

That's interesting, I see what you're saying. I wonder if this is just an oversight, or lack of need to optimize? Not sure.

> Not quite sure what you mean by [] operator.

When I was looking to write the implementation, I got tripped up on the use of iterators. I managed to get most examples in the docs to work using normal collections (Array, Map). For the most part it should be fine. It's just a bit odd since the docs reference ES6 and goja supports 5.1.

I guess overall the big question for me is how closely do we want the API's implementation to be similar to NodeJS's implementation. I was about to change the backing object to URL and URLSearchParams to adhere to a strict ordering of query params.

dop251 commented:
It should be as close as practically achievable. As it's not that much code and there is a reference implementation in javascript, it shouldn't be too difficult to implement it "as-is". See my comment here: https://github.com/dop251/goja_nodejs/pull/40#issuecomment-1337135457, I think you should follow this route, but instead of using url.Values you should have a different implementation which is a ported version from the reference code.

etiennemartin commented:
> See my comment here: https://github.com/dop251/goja_nodejs/pull/40#issuecomment-1337135457, I think you should follow this route

I have an implementation in the works right now with something like this. I'll post an update to this PR once I've done the work.

etiennemartin commented:
I've pushed a pretty big change that changes the backing structure. This brings us closer to the native implementation and allows us to maintain the order of the query parameters. I just need to look into some encoding oddities in the `URL` class right now (See commented out tests).

dop251 commented:
What's the reason for using own structure instead of `url.URL`? I thought it worked quite well...

dop251 commented:
Just don't use `URL.Query()`. I mean it would make sense if you abandoned net.URL completely, but you seem to be using the parser and then just copying everything into a custom struct which does not make a lot of sense in my opinion.

etiennemartin commented:
I'll make the change back. I think you're right. I'll change the nodeURL struct to be a url.URL along with an array of searchParams. The original idea was to move closer to the implementation that NodeJS had.

dop251 commented on url/urlsearchparams.go:
```diff
@@ -0,0 +1,368 @@
+		if len(call.Arguments) > 0 {
+			p := call.Arguments[0]
+			e := p.Export()
+			switch p.ExportType() {
```
ExportType meant as an optimisation that allows to determine the exported type without doing actual export (which can be quite costly, consider an Object with lots of fields).

Therefore, doing unconditional Export() and then calling ExportType() makes no sense: either do Export() and type switch (and risk wasting resources if the value is not of the right type) or call ExportType() and then only do Export() if it's the right type.

dop251 commented on url/urlsearchparams.go:
```diff
@@ -0,0 +1,368 @@
+	return u
+}
+
+func buildParamsFromObject(o map[string]interface{}) *url.URL {
```
Rather than exporting an Object to map, consider working with *Object directly (using Object.Keys() and Object.Get()). This is more memory efficient and will also solve your problem with the property order.

etiennemartin replied:
Do you have an example somewhere of how to use *Object directly for Maps and Arrays. My solutions all end up using export() in the end. So I'm wondering how to pull values from a Map or an Array (via an iterator) using `getSymbol`.

dop251 replied:
I have added Runtime.ForOf which should simplify iterating over Maps and Arrays. Please give it a try.

etiennemartin replied:
Great! I'll take a look and see what I can cook up.

dop251 commented on url/urlsearchparams.go:
```diff
@@ -0,0 +1,368 @@
+	return u
+}
+
+func buildParamsFromArray(r *goja.Runtime, a []interface{}) *url.URL {
```
Same here, consider working with *Object directly. It's a bit less convenient, but still more memory efficient.

dop251 commented on url/urlsearchparams.go:
```diff
@@ -0,0 +1,368 @@
+	return u
+}
+
+func buildParamsFromMap(r *goja.Runtime, m [][2]interface{}) *url.URL {
```
... and here

dop251 commented on url/nodeurl.go:
```diff
@@ -0,0 +1,105 @@
+}
+
+func newFromURL(u *url.URL) *nodeURL {
+	sp, _ := parseSearchQuery(u.RawQuery)
```
Consider making the parsing and updating lazy. I expect there will be a significant number of cases where the parameters are not needed.

etiennemartin replied:
I missed this in the last pass, I'll take a look and add this in.

etiennemartin commented:
I've updated the code and removed the use of Export() with each execution. This approach also makes use for the [ForOf](https://github.com/dop251/goja_nodejs/pull/54#discussion_r1242136577) implementation you provided. Sorry for the delay I was away for a bit. Let me know what you think.

dop251 commented on url/urlsearchparams.go:
```diff
@@ -0,0 +1,424 @@
+			vals, contained := u.getValues(n)
+			if len(call.Arguments) > 1 {
+				for _, v := range vals {
+					cmp := call.Arguments[1].String()
```
This should be hoisted outside the loop. Even though String() may have side effects, checking with nodejs shows it's only called once.

dop251 commented on url/nodeurl.go:
```diff
@@ -0,0 +1,118 @@
+
+type searchParam struct {
+	name  string
+	value []string
```
I've been going though the code and added some comments, and at some point I realise the most important question is why do you need a slice here? I leave the other comments in as most of them are still relevant, however this one needs to be addressed first.

etiennemartin replied:
Good point, I'll simply this and remove the use of an array. This is a relic from before when I didn't maintain ordering of parameters.

dop251 commented on url/nodeurl.go:
```diff
@@ -0,0 +1,118 @@
+		pair := strings.Split(v, "=")
+		name := pair[0]
+		sp := searchParam{name: name, value: []string{}}
+		if len(pair) > 1 {
```
This doesn't look right. If I read this correctly, the code parses "a=1=2" as ["a", ["1", "2"]]. This is not how the nodejs parser works.
For best compatibility I would recommend directly convert the code from nodejs which I quoted earlier.

dop251 reviewed (changes requested):
Hi. Thanks for continuing working on this and apologies for the delay. I've added some comments, please have a look.

etiennemartin commented on url/urlsearchparams.go:
```diff
@@ -0,0 +1,424 @@
+		name := call.Arguments[0].String()
+		if len(call.Arguments) > 1 {
+			value := call.Arguments[1].String()
+			arr := searchParams{}
```
I've since simplified this approach, but the reason I used approach was to allow for deletion of multiple items in one pass. The Search Params might have have many entries with the name. Which would result in all of them being removed.

dop251 replied:
It still doesn't justify the allocation and the full copy. Even if multiple elements are deleted it's still possible to do it with in-place replacement and with a single pass.

etiennemartin commented on url/urlsearchparams.go:
```diff
@@ -0,0 +1,424 @@
+		u := toURL(r, call.This)
+		name := call.Arguments[0].String()
+		if len(call.Arguments) > 1 {
+			value := call.Arguments[1].String()
```
hmmm, I tried to make this change and I don't see `call.Arguments(1)` as a method I can call. It seems to only give access to the array of arguments themselves. So if I did `call.Arguments[1]` and the length is 1, it panics. Am I missing something here?

dop251 replied:
call.Argument(), not Arguments()

etiennemartin commented on url/nodeurl.go:
```diff
@@ -0,0 +1,118 @@
+}
+
+// Second return value determines if name was found in the search
+func (nu *nodeURL) getValues(name string) ([]string, bool) {
```
You're right, you shouldn't have a searchParam with empty values. This was an attempt to optimize the number of iterations we did over the values. But looking back at it again, I'll make another method to determine if a query name exists and split them up. It'll make this easier to read.

etiennemartin commented on url/nodeurl.go:
```diff
@@ -0,0 +1,118 @@
+	return vals, contained
+}
+
+func parseSearchQuery(query string) (searchParams, error) {
```
I'll remove it, good catch.

etiennemartin commented on url/nodeurl.go:
```diff
@@ -0,0 +1,118 @@
+func (s searchParam) String() string {
+	str := url.QueryEscape(s.name)
+	if len(s.value) > 0 {
+		str = fmt.Sprintf("%s=%s", str, strings.Join(s.value, ","))
```
This is used when we read in values during the creation of a `searchParam` object. In this case we want to read in the values as raw-unencoded values. I'll be removing all encoding here to mimic what Node does.

etiennemartin commented on url/nodeurl.go:
```diff
@@ -0,0 +1,118 @@
+	return str
+}
+
+func (s searchParam) String() string {
```
Oops this was an oversight. That being said I did define the ones for `searchParams` without using a pointer since it's a type alias for `[]searchParam`. I'm not sure if this is bad practice or not. But I wasn't sure how to define functions for an array type.

etiennemartin commented on url/nodeurl.go:
```diff
@@ -0,0 +1,118 @@
+
+	str := url.QueryEscape(sp.name)
+	if len(vals) > 0 {
+		str = fmt.Sprintf("%s=%s", str, strings.Join(vals, ","))
```
I really can't remember why I had done this, and it escaped me now. But You're right this is really odd. I'll revert this back where it splits it out into multiple entries with the same key.

dop251 commented on url/urlsearchparams.go:
```diff
@@ -0,0 +1,410 @@
+func buildParamsFromMap(r *goja.Runtime, v goja.Value) *url.URL {
+	query := searchParams{}
+	o := v.ToObject(r)
+	ex := r.Try(func() {
```
Same here

dop251 commented on url/urlsearchparams.go:
```diff
@@ -0,0 +1,410 @@
+			case reflectTypeString:
+				var str string
+				r.ExportTo(v, &str)
+				u = buildParamsFromString(str)
```
It's easier to do `u = buildParamsFromString(v.String())`

dop251 commented on url/nodeurl.go:
```diff
@@ -0,0 +1,117 @@
+	} else {
+		vals = append(vals, fmt.Sprintf("%s=%s", sp.name, sp.value))
+	}
+	return strings.Join(vals, "&")
```
Why do you need the slice and the `Join()` here?

etiennemartin replied:
We don't need it at all, this was a relic from a previous implementation. I'll simplify this.

dop251 commented on url/urlsearchparams.go:
```diff
@@ -0,0 +1,410 @@
+}
+
+// Currently not supporting the following:
+//   - ctor(iterable): Using function generators
```
Is this still relevant?

etiennemartin replied:
With the current implementation this is still relevant. I don't have any code to handle the use of generator functions. It's treated as an Object that generates an empty object. I would need to find a way to call the generator function and process it's output. Any leads on how I could do that given a `goja.Object` that represents a generator function?

dop251 replied:
The output of a generator function is an iterable for which you can use ForOf. BTW, an array is also an iterable, so you can generalise it for objects:

```go
func buildParamsFromObject(r *goja.Runtime, v goja.Value) *url.URL {
    o := v.ToObject(r)
    if o.GetSymbol(goja.SymIterator) != nil {
        return buildParamsFromIterable(r, o)
    }
    ...
}
```
`buildParamsFromIterable` is pretty much the same as the current `buildParamsFromArray`.

etiennemartin replied:
Awesome! I totally missed the `goja.GetSymbol()` functionality in goja. This is great! I'll make this change and push it up. Thanks for the heads up.

dop251 reviewed (changes requested):
It's getting there, just a few minor things that need to be addressed.

etiennemartin commented on url/nodeurl.go:
```diff
@@ -0,0 +1,117 @@
+}
+
+func (nu *nodeURL) getValues(name string) []string {
+	vals := []string{}
```
Interesting, the reason I had done this was to avoid a crash when we called `len(vals)` where the value is returned. I didn't know that `len(x)` could take nil as a value. Good catch!

etiennemartin commented:
I've updated the code base on your comment, the only one I couldn't answer is the one related to the generator functions. I'm not quite sure how I can run a generator function from Go in a clean manner. I'm looking for a way to detect if the parameter is a generator function or not. Is there support for this using goja? If so can you help by pointing me in the right direction?

dop251 commented on url/urlsearchparams.go:
```diff
@@ -0,0 +1,403 @@
+	return u
+}
+
+func stringFromValue(r *goja.Runtime, v goja.Value) string {
```
Why do you need this? The standard `.String()` should handle all these cases.

etiennemartin replied:
Sounds good, I'll remove this.
